### PR TITLE
chore(tests): Convert to typescript and use Jest

### LINKED ISSUE #187: Rewrite tests in TypeScript and Jest
this is a follow up to https://github.com/octokit/webhooks.js/pull/113

@dominguezcelada do you want to take this one?

### PR DESCRIPTION
**Pull request by wolfy1339**
Fixes #187

### CODE REVIEW

File: test/unit/event-handler-wrap-error-handler-test.ts [hunk 1]
```diff
@@ -1,17 +1,16 @@
 import simple from "simple-mock";
-import { test } from "tap";
-import { wrapErrorHandler } from "../../pkg/dist-src/event-handler/wrap-error-handler";
+import { wrapErrorHandler } from "../../src/event-handler/wrap-error-handler";
 
 test("error thrown in error handler", (t) => {
   t.plan(2);
```
Comment (wolfy1339) [inline test/unit/event-handler-wrap-error-handler-test.ts]: Not sure what this is used for. Is it safe to remove?

File: test/unit/event-handler-wrap-error-handler-test.ts [hunk 2]
```diff
@@ -1,17 +1,16 @@
...
-  t.doesNotThrow(() => {
+  expect(
     wrapErrorHandler(() => {
       throw new Error("oopsydoopsy");
-    }, new Error("oops"));
-  });
+    }, new Error("oops"))
+  ).not.toThrow();
 
   t.ok(messages.find((message) => /FATAL/.test(message)));
```
Comment (wolfy1339) [inline test/unit/event-handler-wrap-error-handler-test.ts]: Not sure what to do with these either

File: test/unit/event-handler-wrap-error-handler-test.ts (hunk 1 above)
Comment (gr2m) [reply test/unit/event-handler-wrap-error-handler-test.ts]: I don't think there is something comparable in Jest. This let's you make sure that all assertions are actually run. For example if you have assertions in  callback and the callback is not called at all, then the tests passes although it should not. So when removing the `t.plan` lines, make sure to check that all assertions are run

File: test/unit/event-handler-wrap-error-handler-test.ts (hunk 2 above)
Comment (gr2m) [reply test/unit/event-handler-wrap-error-handler-test.ts]: maybe use [`.toBeTruthy`](https://jestjs.io/docs/en/expect#tobetruthy)?

```suggestion
  expect(messages.find((message) => /FATAL/.test(message))).toBeTruthy();
```

File: test/unit/event-handler-wrap-error-handler-test.ts (hunk 1 above)
Comment (oscard0m) [reply test/unit/event-handler-wrap-error-handler-test.ts]: Maybe this does something similar?
https://jestjs.io/docs/en/expect#expectassertionsnumber

**Comment by oscard0m** [comment]
Would you agree on putting in place ESlint rules for jest? What about the [recommended preset](https://www.npmjs.com/package/eslint-plugin-jest#recommended)?

**Comment by oscard0m** [comment]
@wolfy1339 if you need help or want to split some tests together (as branch from your branch) let me know :)

File: test/unit/event-handler-wrap-error-handler-test.ts (hunk 1 above)
Comment (gr2m) [reply test/unit/event-handler-wrap-error-handler-test.ts]: oh yes, great find, thanks! Let's use that instead

**Comment by wolfy1339** [comment]
> Would you agree on putting in place ESlint rules for jest? What about the [recommended preset](https://www.npmjs.com/package/eslint-plugin-jest#recommended)?

ESLint isn't used (at least exposed to the user), since Prettier is used

**Comment by wolfy1339** [comment]
I need some help. I've got things mostly done here.

The tests in `test/integration/event-handler-test.ts` are failing due to async timeout.
There's also `test/integration/middleware-test.ts`, ` test/unit/middleware-test.ts`, `test/integration/server-test.ts` that need to be looked at, it's mostly just type problems.

<details><summary>Relevant logs:</summary>

```
FAIL  test/unit/middleware-test.ts
  ● Test suite failed to run

    test/unit/middleware-test.ts:12:21 - error TS2345: Argument of type '{ method: string; url: string; }' is not assignable to parameter of type 'IncomingMessage'.
      Type '{ method: string; url: string; }' is missing the following properties from type 'IncomingMessage': aborted, httpVersion, httpVersionMajor, httpVersionMinor, and 45 more.

    12   middleware(state, { method: "POST", url: "/foo" }, {}, next);
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 FAIL  test/integration/middleware-test.ts
  ● Test suite failed to run

    test/integration/middleware-test.ts:32:14 - error TS2345: Argument of type '{ method: string; headers: { "x-github-delivery": string; "x-github-event": string; "x-hub-signature": string; }; url: string; addListener(event: string | symbol, listener: (...args: any[]) => void): EventEmitter; ... 13 more ...; eventNames(): (string | symbol)[]; }' is not assignable to parameter of type 'IncomingMessage'.
      Type '{ method: string; headers: { "x-github-delivery": string; "x-github-event": string; "x-hub-signature": string; }; url: string; addListener(event: string | symbol, listener: (...args: any[]) => void): EventEmitter; ... 13 more ...; eventNames(): (string | symbol)[]; }' is missing the following properties from type 'IncomingMessage': aborted, httpVersion, httpVersionMajor, httpVersionMinor, and 29 more.

    32   middleware(requestMock, responseMock).then(() => {
                    ~~~~~~~~~~~
    test/integration/middleware-test.ts:50:17 - error TS2554: Expected 1 arguments, but got 0.

    50     end: simple.spy(),
                       ~~~~~

      node_modules/@types/simple-mock/index.d.ts:42:12
        42     spy<T>(fn: Fn<T>): Spy<T>;
                      ~~~~~~~~~
        An argument for 'fn' was not provided.
    test/integration/middleware-test.ts:55:14 - error TS2345: Argument of type '{ method: string; headers: { "x-github-delivery": string; "x-github-event": string; "x-hub-signature": string; }; url: string; addListener(event: string | symbol, listener: (...args: any[]) => void): EventEmitter; ... 13 more ...; eventNames(): (string | symbol)[]; }' is not assignable to parameter of type 'IncomingMessage'.
      Type '{ method: string; headers: { "x-github-delivery": string; "x-github-event": string; "x-hub-signature": string; }; url: string; addListener(event: string | symbol, listener: (...args: any[]) => void): EventEmitter; ... 13 more ...; eventNames(): (string | symbol)[]; }' is missing the following properties from type 'IncomingMessage': aborted, httpVersion, httpVersionMajor, httpVersionMinor, and 29 more.

    55   middleware(requestMock, responseMock).then(() => {
                    ~~~~~~~~~~~

 FAIL  test/integration/server-test.ts
  ● Test suite failed to run

    test/integration/server-test.ts:101:11 - error TS2339: Property 'body' does not exist on type 'IncomingMessage'.

    101       req.body = JSON.parse(Buffer.concat(dataChunks).toString());
                  ~~~~
    test/integration/server-test.ts:147:19 - error TS2345: Argument of type 'Spy<unknown>' is not assignable to parameter of type '(event: WebhookEventHandlerError) => void | Promise<void>'.
      Type 'unknown' is not assignable to type 'void | Promise<void>'.
        Type '{}' is missing the following properties from type 'Promise<void>': then, catch, [Symbol.toStringTag], finally

    147   api.on("error", errorHandler);
                          ~~~~~~~~~~~~
    test/integration/server-test.ts:181:31 - error TS2554: Expected 1 arguments, but got 0.

    181   const errorHandler = simple.spy();
                                      ~~~~~

      node_modules/@types/simple-mock/index.d.ts:42:12
        42     spy<T>(fn: Fn<T>): Spy<T>;
                      ~~~~~~~~~
        An argument for 'fn' was not provided.
    test/integration/server-test.ts:182:19 - error TS2345: Argument of type 'Spy<unknown>' is not assignable to parameter of type '(event: WebhookEventHandlerError) => void | Promise<void>'.

    182   api.on("error", errorHandler);
                          ~~~~~~~~~~~~

 FAIL  test/integration/event-handler-test.ts (19.016 s)
  ● events

    : Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:

      3 | import installationCreatedPayload from "../fixtures/installation-created-payload.json";
      4 | 
    > 5 | test("events", (t) => {
        | ^
      6 |   // expect.assertions(7);
      7 | 
      8 |   const eventHandler = createEventHandler({});

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Object.<anonymous> (test/integration/event-handler-test.ts:5:1)

  ● options.transform

    : Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:

      101 | });
      102 | 
    > 103 | test("options.transform", (t) => {
          | ^
      104 |   expect.assertions(2);
      105 | 
      106 |   const eventHandler = createEventHandler({

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Object.<anonymous> (test/integration/event-handler-test.ts:103:1)

  ● multiple errors in same event handler

    : Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:

      141 | });
      142 | 
    > 143 | test("multiple errors in same event handler", (t) => {
          | ^
      144 |   expect.assertions(2);
      145 | 
      146 |   const eventHandler = createEventHandler({});

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Object.<anonymous> (test/integration/event-handler-test.ts:143:1)
```
</details>

**Comment by wolfy1339** [comment]
@dominguezcelada Care to help me on this? I noted the errors I'm getting in my previous comment

**Comment by oscard0m** [comment]
> The tests in `test/integration/event-handler-test.ts` are failing due to async timeout.

I opened a PR to this branch which I think is solving the issue with timeouts: https://github.com/octokit/webhooks.js/pull/195

If you think this is the solution, feel free to merge it in your branch :)

**Comment by oscard0m** [comment]
> There's also `test/integration/middleware-test.ts`, 

https://github.com/octokit/webhooks.js/pull/196

> ` test/unit/middleware-test.ts`,

https://github.com/octokit/webhooks.js/pull/197

> `test/integration/server-test.ts` that need to be looked at, it's mostly just type problems.

https://github.com/octokit/webhooks.js/pull/198 WIP

---

Once we finish this, if this is the right approach to follow, what do you think about getting rid of `simple-mock` library? I think the usage we are giving to this library is already covered by `jest` itself

**Comment by oscard0m** [comment]
btw, before doing merge, I think we should make the first commit (or the resulting squashed one to be `test(<scope>): commit message` as [type](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#type) instead of `chore(<scope>): commit message` 😎

**Comment by oscard0m** [comment]
> > `test/integration/server-test.ts` that need to be looked at, it's mostly just type problems.
> 
> #198 WIP

https://github.com/octokit/webhooks.js/pull/198 ✅  @wolfy1339 

After `tsconfig.json` changes in master with `strict: true` new TypeScript errors were introduced 😢 . Once we validate my PR in here I can jump to fix them if you want

**Comment by gr2m** [comment]
I covered the 9s timeout feature with an integration test here: https://github.com/octokit/webhooks.js/pull/190/files#diff-22ad385c806b5296e5515633b02bd064R256-R302

I added that after you started working on the migration to Jest/TypeScript tests, can you migrate that tests?

I always prefer integration tests over unit tests. If the integration test has the coverage, I'd suggest we remove the unit test

> I'm using Jest's Timer Mocks utility

Looks neat! We can use that in the integration test, too.

**Comment by wolfy1339** [comment]
This is up to date with the master branch. It's been migrated, but it doesn't cover all the code

**Comment by gr2m** [comment]
Got it, I see what's going on. @dominguezcelada tests look good, we are at 100% coverage now.

File: src/sign/index.ts [hunk 3]
```diff
@@ -1,10 +1,6 @@
 import { createHmac } from "crypto";
 
 export function sign(secret: string, payload: string | object): string {
-  if (!secret || !payload) {
-    throw new TypeError("secret & payload required");
-  }
-
```
Comment (gr2m) [inline src/sign/index.ts]: I think we should leave this in. Removing it is a breaking change, albeit a small one. Maybe add above

```ts
// @ts-ignore throw friendly error message when required options are missing
```

**Review by gr2m — CHANGES_REQUESTED** [review]
Thank you both for working through all the tests! I have just one minor comment

File: src/sign/index.ts (hunk 3 above)
Comment (wolfy1339) [reply src/sign/index.ts]: Done.

**Comment by oscard0m** [comment]
ℹ️ I'm taking package-lock.json git conflict  with `master` @wolfy1339 @gr2m

**Comment by oscard0m** [comment]
> ℹ️ I'm taking package-lock.json git conflict with `master` @wolfy1339 @gr2m

@wolfy1339 I'm having issues trying to solve package-lock conflicts. When you are available could you reset the HEAD of your branch with my latest commits and try to solve the package-lock.json conflicts with master? I'm struggling trying to solve them myself I don't know why.

Commits to remove:
[`9dfba7e` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/9dfba7ee37f7c4546ddb683fb7dbfa8a73a94b5a)
[`1ae9147` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/1ae914777ad20799963f8060d2d9ef6895b56a72)
[`1bb18eb` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/1bb18eb3c106e7845b973ebaf63b0b3f37a43991)
[`7c526c2` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/7c526c29d028e05e45bb8b5e360c126c818999b2)
[`9663e8e` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/9663e8e552a33e7e9a056a21de64e5355f00138c)
[`f2716c1` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/f2716c1472e1f3e5f904f73d52700189fbbaf52a)

Sorry for the mess

---

In the meantime I'm working on this PR because there are still some TypeScript issues: https://github.com/octokit/webhooks.js/pull/225

image

**Comment by wolfy1339** [comment]
I use `npm-merge-driver` to automatically resolve `package-lock.json` conflicts.

I also use rebase and not merge

**Comment by oscard0m** [comment]
> I use `npm-merge-driver` to automatically resolve package-locl.json conflicts.
> 
> I also use rebase and not merge

Works like a charm this tool! Thank you so much.

I just merged the problems with sign-test.ts but I've not been able to resolve conflicts from package-lock with master (merging a branch rebased with master or a branch with master merged). I don't understand the issue here

@wolfy1339

**Comment by wolfy1339** [comment]
I'll fix it when I get back home.

**Comment by wolfy1339** [comment]
For reference @dominguezcelada this is how you fix the bad merges & rebase the branch ontop of master
```
git reset HEAD~10 # Revert HEAD to before the commits
git reset --hard # Remove all the changes to files
git rebase origin/master # Rebase the branch on top of master
```

**Comment by wolfy1339** [comment]
@gr2m :cry: You introduced TypeScript errors on master with the `AggregateError` changes in b359fdebf2f6c74b5cfc07189cf735abea8cd252

**Comment by gr2m** [comment]
@wolfy1339 I don't see no errors on master? What's the error you see? How can I reproduce it?

**Comment by wolfy1339** [comment]
I just ran `npm test` and pika spit out a TS error
```
src/event-handler/receive.ts(81,19): error TS2352: Conversion of type 'AggregateError' to type 'WebhookEventHandlerError' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'AggregateError' is missing the following properties from type 'WebhookEventHandlerError': event, errors
src/event-handler/receive.ts(87,66): error TS2345: Argument of type 'WebhookEventHandlerError' is not assignable to parameter of type 'Error'.
  Type 'WebhookEventHandlerError' is missing the following properties from type 'Error': name, message
src/middleware/middleware.ts(83,26): error TS2571: Object is of type 'unknown'.
src/middleware/middleware.ts(83,37): error TS2769: No overload matches this call.
  Overload 1 of 4, '(iterable: Iterable<unknown> | ArrayLike<unknown>): unknown[]', gave the following error.
    Argument of type 'WebhookEventHandlerError' is not assignable to parameter of type 'Iterable<unknown> | ArrayLike<unknown>'.
      Property 'length' is missing in type 'WebhookEventHandlerError' but required in type 'ArrayLike<unknown>'.
  Overload 2 of 4, '(arrayLike: ArrayLike<unknown>): unknown[]', gave the following error.
    Argument of type 'WebhookEventHandlerError' is not assignable to parameter of type 'ArrayLike<unknown>'.
src/types.ts(42,51): error TS2315: Type 'AggregateError' is not generic.

```
Some dependency changed in this branch and introduced this error, but using the same dependencies from this branch, I can reproduce on master

**Comment by gr2m** [comment]
As sorry, I thought you meant the error was on master.

I got a pull request merged into `aggregate-error`: https://github.com/sindresorhus/aggregate-error/pull/13. But before this was merged, I added new the type definitions directly into `@octokit/webhooks`. With the latest change, I upgraded to latest `aggregate-error` and removed to temporary types that I added, as they are now obsolete.

I can look into the error on `wolfy1339-patch-1` tomorrow

**Comment by oscard0m** [comment]
> As sorry, I thought you meant the error was on master.
> 
> I got a pull request merged into `aggregate-error`: [sindresorhus/aggregate-error#13](https://github.com/sindresorhus/aggregate-error/pull/13). But before this was merged, I added new the type definitions directly into `@octokit/webhooks`. With the latest change, I upgraded to latest `aggregate-error` and removed to temporary types that I added, as they are now obsolete.
> 
> I can look into the error on `wolfy1339-patch-1` tomorrow

@wolfy1339 @gr2m Solved [`10386fa` (#188)](https://github.com/octokit/webhooks.js/pull/188/commits/10386fae2eb9d187d511fb8b07255b6999c93778) by upgrading `aggregate-error` package to latest version with @gr2m 's fix (https://github.com/sindresorhus/aggregate-error/pull/13)

**Comment by oscard0m** [comment]
I think we are ready to go now 💪 🍾

File: src/sign/index.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { createHmac } from "crypto";
 
-export function sign(secret: string, payload: string | object): string {
+export function sign(secret?: string, payload?: string | object): string {
```
Comment (gr2m) [inline src/sign/index.ts]: Can you revert this change?

**Review by gr2m — CHANGES_REQUESTED** [review]

File: src/sign/index.ts (hunk 4 above)
Comment (oscard0m) [reply src/sign/index.ts]: > Can you revert this change?

@gr2m without this change we are not able to give coverage to this https://github.com/octokit/webhooks.js/blob/d4bd29b5718fc1390e9b0c33a7df2f340b98f721/src/sign/index.ts#L4-L6

Maybe you prefer this approach on for tests so we give coverage but we bypass the TypeScript check? https://github.com/octokit/webhooks.js/pull/237

**Review by gr2m — APPROVED** [review]
Outstanding work @wolfy1339 & @dominguezcelada 🥇🥇🎊
